Mathlib/Analysis/VonNeumannAlgebra/Basic.lean

```diff
@@ -133,4 +134,34 @@ theorem mem_commutant_iff {S : VonNeumannAlgebra H} {z : H →L[ℂ] H} :
 theorem commutant_commutant (S : VonNeumannAlgebra H) : S.commutant.commutant = S :=
   SetLike.coe_injective <| by simp
 
+open LinearMap in
+/-- An idempotent operator `e` is an element in the von Neumann algebra `S`
+if and only if `range e` and `ker e` are `S.commutant` invariant subspaces. -/
+theorem mem_iff_range_and_ker_mem_invtSubmodule_commutant_of_isIdempotentElem {e : H →L[ℂ] H}
+    (h : IsIdempotentElem e) (S : VonNeumannAlgebra H) :
+    e ∈ S ↔ ∀ y ∈ S.commutant,
+    range e ∈ Module.End.invtSubmodule y ∧ ker e ∈ Module.End.invtSubmodule y := by
```

Comment:
You should factor this out into several lemmas, which will be easy using some results in your recent PRs.

1. `e ∘L y ∘L e = y ∘L e ↔ range e ∈ invtSubmodule y`
2. `e * y * (1 - e) = 0 ↔ ker e ∈ invtSubmodule y`
3. `Commute e y ↔ range e ∈ invtSubmodule y ∧ ker e ∈ invtSubmodule y`

Then proving this result should be quite easy.